File: leetcode-04/solutions/python3/0030.py
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        if not s or not words:
            return []

        m = len(s)
        n = len(words[0])

        ans = []

        dict = collections.Counter(words)

        for i in range(n):
            index = i
            count = 0
            tempDict = collections.defaultdict(int)
            for j in range(i, m - n + 1, n):
                str = s[j: j + n]
                if str in dict.keys():
                    tempDict[str] += 1
                    count += 1
                    while tempDict[str] > dict[str]:
                        tempDict[s[index: index + n]] -= 1
                        count -= 1
                        index += n
                    if count == len(words):
                        ans.append(index)
                        tempDict[s[index: index + n]] -= 1
                        count -= 1
                        index += n
                else:
                    tempDict.clear()
                    count = 0
                    index = j + n

        return ans

File: leetcode-04/solutions/cpp/0030.cpp
class Solution {
 public:
  vector<int> findSubstring(string s, vector<string>& words) {
    if (s.empty() || words.empty()) return {};

    const int m = s.length();
    const int n = words[0].length();

    vector<int> ans;

    unordered_map<string, int> map;
    for (string& word : words) map[word]++;

    for (int i = 0; i < n; i++) {
      int index = i;
      int count = 0;
      unordered_map<string, int> tempMap;
      for (int j = i; j <= m - n; j += n) {
        string str = s.substr(j, n);
        if (map.count(str)) {
          tempMap[str]++;
          count++;
          while (tempMap[str] > map[str]) {
            tempMap[s.substr(index, n)]--;
            count--;
            index += n;
          }
          if (count == words.size()) {
            ans.push_back(index);
            tempMap[s.substr(index, n)]--;
            count--;
            index += n;
          }
        } else {
          tempMap.clear();
          count = 0;
          index = j + n;
        }
      }
    }

    return ans;
  }
};